### PR TITLE
Optimize getting config item values 

### LINKED ISSUE #12504: ConfigNamespace is slow
In investigating some performance issues for #12491 I found that a not-insignificant contributor is the performance of looking up values in `ConfigNamespace`.

It is actually quite slow if you look at the code, specifically for `ConfigItem.__get__`, which goes through quite a lot of steps of looking up the correct config object, reading the value from it, checking aliases, validating the value, etc.

This means that if some config option is used in a tight loop or in other common code paths it can be a bit of a drag.

For astropy.io.fits this comes up in particular for header parsing, for which there are a couple of different config options, and these get checked for *every single card* in a header when parsing its value (not a huge problem if you don't read every card in the header, but if you read a few hundred cards of each header it blows up).

I'm still analyzing exactly how much it contributes to the overall time of parsing headers, but it seems to be a significant component.  I have some ideas for fixing this but still experimenting.

### PR DESCRIPTION
**Pull request by saimn**
Fix #12504 (I think, doing better might be possible but not sure without breaking some feature of our complicated configuration system):

As explained there, getting a config item is quite slow, values being stored within the ConfigObj object . And io.fits accesses config items when e.g. accessing the value of a card, this means a lot of config item calls for tables especially with many columns. 

So here I'm adding a cache on the ConfigItem descriptor, allowing to get the key much faster while still storing its value on the ConfigObj object.

### Description

This pull request is to address ...

Fixes #<Issue Number>

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by saimn** [comment]
Forgot to add some timings.
Before:
```
In [2]: %timeit fits.conf.enable_record_valued_keyword_cards
2.69 μs ± 502 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [3]: %timeit fits.getdata("astropy/io/fits/tests/data/tb.fits")
1.73 ms ± 338 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

After:
```
In [2]: %timeit fits.conf.enable_record_valued_keyword_cards
55.2 ns ± 0.163 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [3]: %timeit fits.getdata("astropy/io/fits/tests/data/tb.fits")
1.09 ms ± 6.35 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

File: astropy/config/configuration.py [hunk 1]
```diff
@@ -327,13 +317,22 @@ def __init__(
...
         if obj is None:
             return self
-        return self()
+        # cache value on the descriptor itself, to avoid repeated accesses
+        # to the ConfigObj object which is much slower
+        try:
+            val = self.value
+        except AttributeError:
+            val = self.value = self()
+        return val
```
Comment (pllim) [inline astropy/config/configuration.py:335]: What about this?

```suggestion
        if not hasattr(self, "value"):
            self.value = self()
        return self.value
```

File: astropy/config/configuration.py [hunk 2]
```diff
@@ -402,7 +403,11 @@ def reload(self):
             The new value loaded from the configuration file.
 
         """
-        self.set(self.defaultvalue)
+        try:
+            del self.value
+        except AttributeError:
+            pass
```
Comment (pllim) [inline astropy/config/configuration.py:409]: Maybe like this instead? Though I don't see how this is equivalent to `self.set(self.defaultvalue)`.

```suggestion
        if hasattr(self, "value"):
            del self.value
```

File: docs/changes/config/17461.perf.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Optimize getting config item values.
```
Comment (pllim) [inline docs/changes/config/17461.perf.rst:1]: Can we roughly quantify this?

File: astropy/config/configuration.py (hunk 1 above)
Comment (mhvk) [reply astropy/config/configuration.py:335]: `try/except` has no cost if there is no exception, so I think it is good to use it (`hasattr` uses it internally, but now you do have the cost of the `if` statement). I'd try to shave off another few ns with
```
try:
    return self.value
except AttributeError:
    self.value = val = self()
    return val
```

File: astropy/config/configuration.py [hunk 4]
```diff
@@ -360,9 +359,11 @@ def set(self, value):
                 f" {e.args[0]}"
             )
 
+        # store value on the ConfigObj instance...
         sec = get_config(self.module, rootname=self.rootname)
-
         sec[self.name] = value
+        # and on the descriptor
+        self.value = value
```
Comment (mhvk) [inline astropy/config/configuration.py:366]: While you are at it, I think inside `set_temp` below, `self.set(value)` should really be inside the `try/finally` (if it raises an exception, the value should be unset).

**Review by mhvk — APPROVED** [review]
This looks good to me, though I haven't really looked at this code before, so it would be good for someone else to check too. Benchmarks certainly look good (the one bad one should be an unrelated fluke). But need some precommit fix or so.

File: astropy/config/configuration.py (hunk 4 above)
Comment (saimn) [reply astropy/config/configuration.py:365]: done

File: astropy/config/configuration.py (hunk 2 above)
Comment (saimn) [reply astropy/config/configuration.py:408]: For `self.set(self.defaultvalue)` it's annoying but I don't remember :grimacing: (I really should open PRs faster!)

File: docs/changes/config/17461.perf.rst (hunk 3 above)
Comment (saimn) [reply docs/changes/config/17461.perf.rst:1]: It's not a noticeable change if you get a few config values but it is for `io.fits` which uses config items when e.g. getting card values. So when reading many keywords or a table with many columns it starts to matter.

File: astropy/config/configuration.py (hunk 1 above)
Comment (saimn) [reply astropy/config/configuration.py]: yeah try/except is commonly used in such cases both for perf and also `hasattr` can have some side effects so simpler is better here.

File: astropy/config/configuration.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/config/configuration.py:408]: note that the `try/except` form is more robust to parallelism (which *might* matter in the wake of free-threaded Python ?)

Comment (pllim) [reply astropy/config/configuration.py:408]: As long as hitting `except` is rare...

File: docs/changes/config/17461.perf.rst (hunk 3 above)
Comment (pllim) [reply docs/changes/config/17461.perf.rst:1]: I didn't know FITS uses this... 🙈

**Review by pllim — APPROVED** [review]
Thanks!

**Comment by mhvk** [comment]
Unfortunately, this caching of config items introduced a bug, in that `reload_config` does not reset the values cached here - see #18138.

**Pull request merged**